### PR TITLE
Misc Improvements + respecting post visibility state when listing threadmarks

### PR DESCRIPTION
**Pull request by Xon**
- Use code listener event hints
- On xf_thread, has_threadmark -> threadmark_count
- Improved SQL for updating, hook DataWriter to ensure when a post's visibility state changes the threadmark model is notified.
- Attempted to fixed up install process to apply incremental updates.

### CODE REVIEW

File: upload/library/Sidane/Threadmarks/Model/Threadmarks.php [hunk 1]
```diff
@@ -3,59 +3,107 @@
...
+    {
+        $db = $this->_getDb();
+        
+        // remove orphaned threadmarks
+        $db->query('
+            DELETE `threadmarks`
+            FROM `threadmarks`
+            LEFT JOIN xf_post AS post on post.thread_id = threadmarks.thread_id and post.post_id = threadmarks.post_id
+            where `threadmarks`.thread_id = ? 
+        ', $thread_id);
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/Model/Threadmarks.php]: This query deleted every existing threadmark record when it ran during the add-on upgrade.

File: upload/library/Sidane/Threadmarks/Install.php [hunk 2]
```diff
@@ -2,37 +2,50 @@
...
-        thread_id INT UNSIGNED NOT NULL,
-        post_id INT UNSIGNED NOT NULL,
-        label VARCHAR(100) NOT NULL,
-        KEY thread_id (thread_id)
-      ) ENGINE = InnoDB CHARACTER SET utf8 COLLATE utf8_general_ci
-    ");
+        if ($version == 1)
+        {
+            $db->query("ALTER TABLE xf_thread DROP COLUMN has_threadmarks");
+            $db->query("ALTER TABLE xf_thread DROP INDEX thread_id ");
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/Install.php]: I don't think this should be here. Add-on upgrade broke when it reached this line.

File: upload/library/Sidane/Threadmarks/Install.php [hunk 3]
```diff
@@ -2,37 +2,50 @@
...
-    try
-    {
-      $db->query("ALTER TABLE xf_thread ADD COLUMN has_threadmarks INT UNSIGNED DEFAULT 0 NOT NULL AFTER prefix_id");
+        if ($version < 2)
+        {            
+            $db->query("ALTER TABLE xf_thread ADD COLUMN threadmark_count INT UNSIGNED DEFAULT 0 NOT NULL");
+            if (!$tables_created)
+            {
+                $db->query("ALTER TABLE threadmarks add unique index thread_post_id (`thread_id`,`post_id`)");
+                $db->query("ALTER TABLE threadmarks add unique index thread_post_id (`thread_id`,`post_id`)");
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/Install.php]: Duplicate of previous line. Breaks add-on upgrade.

File: upload/library/Sidane/Threadmarks/Model/Threadmarks.php [hunk 4]
```diff
@@ -3,59 +3,107 @@
 class Sidane_Threadmarks_Model_Threadmarks extends XenForo_Model
 {
 
-  public function createThreadMark($thread, $post, $label) {
+  public function SetThreadMark($thread_id, $post_id, $label) {
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/Model/Threadmarks.php]: Inconsistent method name, should be `setThreadMark`.

**Comment by sidane** [comment]
Thanks for the improvements. A few blockers in my previous comments that need fixing. 

I want to look at the callbacks in the datawriter in more detail and test these changes further locally. Once the add-on upgrades smoothly I'll take another look.

Also, please remove all tabs (e.g. `Cache.php`) and use 2 space indentation throughout, I see a number of places where you're using 4.

Cheers.

File: upload/library/Sidane/Threadmarks/Deferred/Cache.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+		
+		return $data;        
+    }
+
+
+	public function canCancel()
+    {
+		return true;
+	}    
+}
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/Deferred/Cache.php:42]: Formatting is a bit all over the place in this file. Mixture of tabs and spaces as well - please remove all tabs.

_Formatting nazi alert_ :wink:

File: upload/library/Sidane/Threadmarks/DataWriter/Post.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            $decrementCount = ($this->isUpdate() && $this->getExisting('message_state') == 'visible');
+            $this->_getThreadMarkModel()->deleteThreadMark($threadmark['threadmark_id'], $decrementCount);
+        }
+	}    
+    
+    protected function _getThreadMarkModel()
+    {
+        return $this->getModelFromCache('Sidane_Threadmarks_Model_Threadmarks');
+    }
+}
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/DataWriter/Post.php:43]: Formatting a bit ugly again - sorry!

File: upload/library/Sidane/Threadmarks/Install.php (hunk 2 above)
Comment (Xon) [inline upload/library/Sidane/Threadmarks/Install.php]: you are right that should be:
ALTER TABLE threadmarks DROP INDEX thread_id

File: upload/library/Sidane/Threadmarks/Model/Threadmarks.php (hunk 1 above)
Comment (Xon) [inline upload/library/Sidane/Threadmarks/Model/Threadmarks.php]: This was missing a where clause. should have a "and post.post_id is null"

File: upload/library/Sidane/Threadmarks/Deferred/Cache.php (hunk 5 above)
Comment (Xon) [inline upload/library/Sidane/Threadmarks/Deferred/Cache.php:42]: My damn editor wasn't configured to convert tabs to spaces. Ugh. hate tab character.

**Comment by Xon** [comment]
I've hopefully addressed the issues you've raised with the upgrade process.

**Comment by sidane** [comment]
Thanks, I'll take a look again this evening.

File: upload/library/Sidane/Threadmarks/DataWriter/Post.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        else if ($this->isUpdate() && $this->getExisting('message_state') == 'visible')
+        {
+          $this->_getThreadMarkModel()->modifyThreadMarkCount($this->get('thread_id'), -1);
+        }
+      }
+    }        
+  }
+    
+  protected function _messagePostDelete()
+  {
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/DataWriter/Post.php:28]: I think this method needs to call `parent::_messagePostDelete();` as it extends `XenForo_DataWriter_DiscussionMessage_Post` which implements its own `_messagePostDelete` method.

Comment (sidane) [inline upload/library/Sidane/Threadmarks/DataWriter/Post.php:28]: Also, this method doesn't seem to be called at all. I've tested it locally by soft deleting and permanently deleting a post with a threadmark and the counter never decrements. Also tried raising an error in the method and it's never raised.

File: upload/library/Sidane/Threadmarks/ControllerPublic/Post.php [hunk 8]
```diff
@@ -33,12 +33,12 @@ public function actionThreadmark()
           $phrase = 'threadmark_deleted';
           XenForo_Model_Log::logModeratorAction('post', $post, 'delete_threadmark', array(), $thread);
         } else {
-          $threadmarksModel->updateThreadmark($existingThreadmark, $label);
+          $threadmarksModel->SetThreadMark($thread['thread_id'], $post['post_id'], $label);
```
Comment (sidane) [inline upload/library/Sidane/Threadmarks/ControllerPublic/Post.php]: So this works even though the method is defined as `setThreadMark` but called here with a capital S?

Please rename it to `setThreadMark` anyway for consistency (and on line line 41 as well).

File: upload/library/Sidane/Threadmarks/DataWriter/Post.php (hunk 7 above)
Comment (sidane) [inline upload/library/Sidane/Threadmarks/DataWriter/Post.php:28]: Trying it again the counter is decrementing now but my debug code in the method still isn't raising, I need to take another look at it.

**Comment by sidane** [comment]
A fresh upgrade worked with your changes :+1:  I still want to do some more testing though.

There's also still a few files that need consistent formatting:
- Install.php
- Listener.php
- Models/Threadmarks.php

Thanks.

File: upload/library/Sidane/Threadmarks/ControllerPublic/Post.php (hunk 8 above)
Comment (Xon) [inline upload/library/Sidane/Threadmarks/ControllerPublic/Post.php]: I thought I had tested that, and that shouldn't work =\

File: upload/library/Sidane/Threadmarks/DataWriter/Post.php (hunk 7 above)
Comment (Xon) [inline upload/library/Sidane/Threadmarks/DataWriter/Post.php:28]: Good catch on the missing parent::_messagePostDelete call.

This call chain is triggered on a permanently deleting a post:
XenForo_DataWriter_DiscussionMessage::_postDelete
XenForo_DataWriter_DiscussionMessage::_messagePostDelete

**Comment by Xon** [comment]
OK, hopefully this round works better :)

**Comment by sidane** [comment]
Testing this again from scratch, the `_messagePostDelete` callback doesn't seem to be called.
- I have a thread with 3 threadmarks. 
- `threadmark_count` on `xf_thread` is correctly set to 3.
- I permanently delete the post but `threadmark_count` remains at 3 and the threadmark is still in the `threadmarks` table.

Putting `var_dump('foo')` at the start of `_messagePostDelete` raises no errors when I delete a post.

Not sure what the issue is, are you able to replicate this @Xon?

**Comment by Xon** [comment]
I edited my local copy with XenForo_Error::debug() as it logs stack traces, like so:

```
  protected function _messagePostDelete()
  {
  XenForo_Error::debug("before parent: ".$this->get('thread_id')." ". $this->get('post_id'));
    parent::_messagePostDelete();
XenForo_Error::debug("after parent");
    $threadmark = $this->_getThreadMarkModel()->getByThreadIdAndPostId($this->get('thread_id'), $this->get('post_id'));
    XenForo_Error::debug("after parent". var_export($threadmark, true));
    if (!empty($threadmark))
    {
      $decrementCount = ($this->isUpdate() && $this->getExisting('message_state') == 'visible');
      $this->_getThreadMarkModel()->deleteThreadMark($threadmark['threadmark_id'], $decrementCount);
    }
  }
```

It produced the following when hard deleting a post:
http://puu.sh/esRMf/cef83dd8a9.png

**Comment by sidane** [comment]
And is it successfully deleting the threadmark / decrementing the counter for you?

**Comment by Xon** [comment]
Nope. Trying to figure out what the hell is going wrong.

**Comment by Xon** [comment]
Well that was quite silly. It is working now for me.

**Comment by sidane** [comment]
Looks good now. I'll deploy these updates to my site for a few days to be sure it's production ready.

In the meantime I'll address those issues you've opened. Good chance for me to learn some XenForo best practices :smile: I'll make pull requests for the changes and would appreciate your feedback on them if you've got time.

Once those issues are sorted, I'll release v1.1 to the XenForo community.

**Pull request merged**